docs/design/sum_types.md

```diff
@@ -0,0 +1,200 @@
+primitive which hasn't been designed yet, but the key to defining a sum type's
+interface is enabling it to support pattern matching. To do that, the sum type
+has to specify two things:
+
+-   The set of all possible alternatives, including their names and parameter
+    types, so that the compiler can typecheck the `match` body, identify any
+    unreachable `case`s, and determine whether any `case`s are missing.
+-   The algorithm that, given a value of the sum type, determines which
+    alternative is present, and specifies the values of its parameters.
+
+Here's how that would look if `Optional` were defined as a class:
+
+```carbon
+class Optional(T:! Type) {
+  // Factory functions
+  fn Some(value: T) -> Self;
+  let None:! Self;
+
+  private var has_value: bool;
+  private var value: T;
+
+  interface MatchContinuation {
+    let template ReturnType:! Type;
+    fn Some[addr me: Self*](value: T) -> ReturnType;
+    fn None[addr me: Self*]() -> ReturnType;
+  }
+
+  external impl as Match(MatchContinuation) {
+    fn Op[me: Self, Continuation:! MatchContinuation](
+        continuation: Continuation*) -> Continuation.ReturnType {
+      if (me.has_value) {
+        return continuation->Some(me.value);
+      } else {
+        return continuation->None();
+      }
+    }
+  }
+
+  // Operations like destruction, copying, assignment, and comparison are
+  // omitted for brevity.
+}
+```
+
+In this code, `Optional` makes itself available for use in pattern matching by
+declaring that it implements the `Match` interface. `Match` takes an interface
+argument, called the _continuation interface_, which specifies the set of
+possible alternatives by declaring a method for each one. In this case, we pass
+`Optional.MatchContinuation` as the continuation interface.
+
+When compiling a `match` statement, the compiler checks that the type being
+matched implements `Match(C)` for some continuation interface `C`. Then, it
+notionally transforms the `match` body into a class that implements `C`, with
+one method for each `case`, and then passes that to `Match.Op` on the object
+being matched. For example, the `match` statement shown earlier might be
+transformed into:
+
+```carbon
+class __MatchStatementImpl {
+  fn Make() -> Self { return {}; }
+
+  impl as Match(Optional.MatchContinuation) where .ReturnType = () {
+    fn Some(the_value: i32) {
+      Print(the_value);
+    }
+    fn None() {
+      Print("None");
+    }
+  }
+}
+
+my_opt.(Match.Op)(__MatchStatementImpl.Make());
+```
+
+Thus, a `match` statement works by invoking the sum type's `Match.Op` method,
+which is responsible for determining which alternative the sum object
+represents, and then invoking the compiler-supplied continuation that
+corresponds to that alternative. In order for this scheme to work, `Match.Op` is
+required to invoke exactly one method of `MatchContinuation`, and to do so
+exactly once.
```

Comment:
I wonder if we can somehow enforce that in the type system. We already have a guarantee that if `Match.Op` returns, it must have invoked a method of the continuation, because there's no other way to construct an instance of the return type. I think what we'd be missing is a way to make the return type non-discardable.

I'm assuming that it's OK for `Match.Op` to, for example, terminate the thread or the program, or simply not terminate, instead of running exactly one continuation exactly once, but an "if it returns" would help here.